mPower2/mPower2/PassiveCollectors.swift

```diff
@@ -225,25 +235,456 @@ class PassiveDisplacementCollector : NSObject, PassiveCollector, CLLocationManag
     func start() {
+        // No point in bothering if we can't tell whether or not they're walking.
+        guard CMMotionActivityManager.isActivityAvailable() else { return }
+        
+        self.activityManager = CMMotionActivityManager()
         
+        // Request permissions when starting (i.e. at launch, or via other user interaction). For this we use a temporary
+        // RSDMotionRecorder instance.
+        let appDelegate = AppDelegate.shared!
+        let viewController = appDelegate.rootViewController ?? appDelegate.window!.rootViewController!
+        let viewModel = PassiveGaitModel(schemaIdentifier: self.schemaIdentifier)
+        let tempRecorder = self.config.instantiateController(with: viewModel) as? RSDMotionRecorder
+        tempRecorder?.requestPermissions(on: viewController, { (action, result, error) in
+            // If we got permission, start listening for activity updates, and if it says we've started walking,
+            // record an up-to-30-second burst of motion sensor data.
+            self.activityManager?.startActivityUpdates(to: OperationQueue.main, withHandler: { (activity) in
+                // If we get called without an activity object, ignore it.
+                guard let activity = activity else { return }
+                
+                // Were we already recording a walk?
+                let weWereWalking = self.recorder != nil
+                
+                // Are we walking now?
+                let weAreWalkingNow = (activity.walking == true) && (activity.confidence == .high)
+                
+                // We only care when the walking state changes:
+                guard weWereWalking != weAreWalkingNow else { return }
+                
+                // If we were recording a walk but we're not walking anymore, cut it short and go with what we've got.
+                if weWereWalking {
+                    #if DEBUG
+                    self.debugNotification(title: "Stopped walking", body: "\(self.describe(activity: activity))")
+                    #endif
+                    self.stopRecorderAndUpload()
+                }
+                // If we were *not* recording a walk already, but now we're walking, record and upload a 30-second burst
+                // of motion sensor data.
+                else {
+                    #if DEBUG
+                    self.debugNotification(title: "Logging", body: "\(self.describe(activity: activity))")
+                    #endif
+                    self.recordMotionSensorBurst()
+                }
+            })
+            
+            // Also, listen for active gait recorders being started, and when that happens shut ours down so we don't interfere.
+            // This will be called on the posting queue, which for this notification is always main.
+            self.listener = NotificationCenter.default.addObserver(forName: .RSDMotionRecorderWillStart, object: nil, queue: nil, using: { (notification) in
```

Comment:
It needs to run synchronously on main so that the old CMMotionManager gets stopped before the new one gets started. Easier (and, I thought, cleaner) to just ensure that the notification gets posted on the main queue than to mess with the whole dispatch-sync-from-main-to-main-without-deadlocking dance.

As for weak self and retain loops, Xcode isn't yelling at me about that, even if I tell it to analyze the code, so I was assuming it's not a problem here. It at least used to warn about retain loops.